lib/ProductOpener/DataQuality.pm

```diff
@@ -0,0 +1,150 @@
+# This file is part of Product Opener.
+#
+# Product Opener
+# Copyright (C) 2011-2019 Association Open Food Facts
+# Contact: contact@openfoodfacts.org
+# Address: 21 rue des Iles, 94100 Saint-Maur des Fossés, France
+#
+# Product Opener is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+=head1 NAME
+
+ProductOpener::DataQuality - check the quality of data for products
```

Comment:
It does not only check quality but it also stores quality tags inside product data isn't it? Why not separate checking and storing data?

Reply:
Yes it does store the results of the quality checks as quality tags. But that's the best way to do it I think, having an extra layer between results and storing the results would only add code and complexity. It is also the way all other processing on product data work: we pass a reference to the product object, and it is enriched with new analysis etc.